File: python-service/app/models/validation.py
"""Pydantic models for validation results and request payloads."""

from typing import List, Literal, Optional

from pydantic import BaseModel

from app.models.address import StructuredAddress


class ValidationResult(BaseModel):
    """Result of validating a single address against USPS."""

    original_address: StructuredAddress
    standardized_address: Optional[StructuredAddress] = None
    status: Literal["valid", "invalid", "invalid_input", "parse_failed"]
    error_message: Optional[str] = None


class ZipcodeCityResult(BaseModel):
    """Result of verifying a zipcode-city match."""

    zipcode: str
    city: str
    status: Literal["match", "mismatch"]
    valid_cities: Optional[List[str]] = None


class ZipcodeCityRequest(BaseModel):
    """Request payload for zipcode-city verification."""

    zipcode: str  # 5-digit or 5+4 format (e.g., "20500" or "20500-0001")
    city: str


class ParseRequest(BaseModel):
    """Request payload for LLM-powered address parsing."""

    raw_address: str


class ParseValidationResult(BaseModel):
    """Result of parsing unstructured text and validating the extracted address."""

    raw_text: str
    parsed_address: Optional[StructuredAddress] = None
    validation_result: Optional[ValidationResult] = None
    status: Literal["valid", "invalid", "parse_failed", "service_unavailable"]
    error_message: Optional[str] = None


class BulkValidationResult(BaseModel):
    """Result of bulk address validation from an Excel file upload."""

    total_rows: int
    results: List[ValidationResult]  # Preserves row order

File: express-service/src/models/validation.ts
import { StructuredAddress } from './address';

/**
 * Result of validating a single address against USPS.
 */
export interface ValidationResult {
  original_address: StructuredAddress;
  standardized_address?: StructuredAddress;
  status: 'valid' | 'invalid' | 'invalid_input' | 'parse_failed';
  error_message?: string;
}

/**
 * Result of verifying a zipcode-city pair.
 */
export interface ZipcodeCityResult {
  zipcode: string;
  city: string;
  status: 'match' | 'mismatch';
  valid_cities?: string[];
}

/**
 * Request body for zipcode-city verification.
 */
export interface ZipcodeCityRequest {
  zipcode: string;
  city: string;
}

/**
 * Request body for unstructured address parsing.
 */
export interface ParseRequest {
  raw_address: string;
}

/**
 * Result of parsing unstructured text and validating the parsed address.
 */
export interface ParseValidationResult {
  raw_text: string;
  parsed_address?: StructuredAddress;
  validation_result?: ValidationResult;
  status: 'valid' | 'invalid' | 'parse_failed' | 'service_unavailable';
  error_message?: string;
}

/**
 * Result of bulk address validation from an Excel file upload.
 */
export interface BulkValidationResult {
  total_rows: number;
  results: ValidationResult[];
}
